### ISSUE TITLE
Define tunable hyperparameters for reservoir computing models

### ISSUE THREAD

**Issue by juanaurre99** [Cyber-Physical-Systems-in-Mech-Eng/pyReCo#68]
## Summary

This issue aims to define a core set of tunable hyperparameters for reservoir computing models in this library. These parameters will be exposed to tuning utilities (e.g., grid search, Bayesian optimization) in a generic and reusable way.

## Tunable Hyperparameters (Initial Set)

The following hyperparameters are proposed for tuning:

- `spectral_radius` (float): Controls the echo state property
- `input_scaling` (float): Scales the input weights
- `leak_rate` (float): Leak rate for leaky ESNs
- `ridge_alpha` (float): Regularization strength for the readout
- `bias_scaling` (float): Bias weight scaling
- `activation` (str): Nonlinearity used in the reservoir (e.g., `tanh`, `relu`)

## Excluded Parameters

As discussed in our last meeting, we are **excluding** the following from the tuning space:

- `reservoir_size`: Changing this would require reinitializing the reservoir
- `density`: Also requires reinitialization due to changes in the reservoir topology

These structural parameters are fixed during model instantiation to preserve consistency and computational efficiency during tuning.

## Next Steps

Once we finalize this list, the next step will be to determine how the `Tuner` class will interface with these hyperparameters — i.e., how to dynamically modify them without breaking the internal model structure. This will be handled in a follow-up issue or in the `feature/hp-tuning` branch.

## Action Items

- [x] Finalize the hyperparameter list
- [ ] Confirm types and valid ranges
- [ ] Prepare for integration with the `Tuner` class

**Tagging for input:** @maneesh51  @mertstend

**Comment by juanaurre99** [comment]
@mertstend: Regarding input_scaling, I’ve read that it is—along with spectral radius and leaking rate—one of the most important hyperparameters to tune in ESNs (see https://www.ai.rug.nl/minds/uploads/PracticalESN.pdf). So I thought it could make sense to include it in the library as a future option.
As for the AI-generated content: I use these tools mainly to help make the issues clearer and more readable, but yes—I’ll make sure to be more careful with the technical details going forward. Thanks again!

**Comment by mertstend** [comment]
as you were asking for the user perspective. I'd suggest the following:

**definition of search parameters** using a dictionary with key (hyperparamer) and value (lower and upper bound, or even grid of values to try out). Of course the method has some default values to always try out (if the user did not overwrite them). 

`param_space = {'alpha': [0.1, 0.9], 'activation':['sigmoid', 'tanh', 'relu']}`

Then the hyperparameter search:
```
hyperp_searcher = HyperParameterSearch(param_space)
best_model, param_study = hyperp_searcher.optimize(model, x, y)
```

best_model is a simple RC model, param_study is a dictionary

**Comment by mertstend** [comment]
Hi, 

thanks for those ideas. I'd like to step back for a second and focus on the main capabilities that this Tuner class should have. Most importantly, it will rely on some methods that are guaranteed to exist in any pyReCo model object: 

```
.set_params(**kwargs) to configure hyperparameters,
.fit(X, y) to train on training data,
.evaluate(X, y, metric) to compute performance on validation data.
```

I think that we currently do not have a generic set_params (maybe call it set_hyperparams) method. This would be the very first step, which needs very serious validation and tracking (e.g. when changing the spectral radius we actually have to re-scale the network, i.e. a change in parameter may trigger subsequent operations). 

**I suggest to work on that generic setter and getter method for a range of hyperparameters first in a separate branch (including unit tests).**

**Comment by mertstend** [comment]
Don't want to push my own ideas too far here, but this would be a minimal implementation that I would suggest:

```
from abc import ABC, abstractmethod
from typing import Dict, List, Any
from itertools import product
import numpy as np


### ---- Base Tuner ----
class Tuner(ABC):
    def __init__(self, model, param_space: Dict[str, List[Any]], metric: str,
                 X_train, y_train, X_val, y_val):
        self.model = model
        self.param_space = param_space
        self.metric = metric
        self.X_train = X_train
        self.y_train = y_train
        self.X_val = X_val
        self.y_val = y_val

    @abstractmethod
    def optimize(self) -> Dict[str, Any]:
        pass

### ---- Grid Search Tuner ----
class GridSearchTuner(Tuner):
    def optimize(self) -> Dict[str, Any]:
        keys = list(self.param_space.keys())
        candidates = [dict(zip(keys, values)) for values in product(*self.param_space.values())]

        best_score = float('-inf')
        best_params = None

        for params in candidates:
            self.model.set_params(**params)
            self.model.fit(self.X_train, self.y_train)
            score = self.model.evaluate(self.X_val, self.y_val, metric=self.metric)

            if score > best_score:
                best_score = score
                best_params = params

        return {'best_params': best_params, 'best_score': best_score}

### ---- Random Search Tuner ----
class RandomSearchTuner(Tuner):
...
```

with the desired usage: 

```
# Define parameter space
param_space = {
    'spectral_radius': [0.1, 0.95],
    'activation': ['sigmoid', 'tanh']
}

# Instantiate tuner
tuner = GridSearchTuner(model, param_space, metric='mse',
                        X_train=X_train, y_train=y_train, X_val=X_val, y_val=y_val)

# Run optimization
result = tuner.optimize()
print(result)
```

**Comment by juanaurre99** [comment]
Hello Merten, I appreciate your feedback and examples. I implemented it this way because I think it’s a good idea to have a more customizable approach to tuning—for instance, if a user wants to minimize a combination of different metrics outside of the model.evaluate method for their own tailored task. But yes, it requires more thought for now. I wanted to provide a first working example of how the implementation could look. Optuna, for example, has some very good features like parallelization and built-in search algorithms.

As for set_params, I’ll work on that. I didn’t include it in the first version since we needed to discuss how to integrate it into the codebase and how to modify hyperparameters from outside. Let me know if you have any more suggestions. Thanks!

**Comment by mertstend** [comment]
@juanaurre99 : thanks! I appreciate your thoughts and ideas. Let's build this step-by-step, and validate fast. Also, this will help us to review your PR and check implementations quickly. So please keep it simple for the start, and then we take iterations. For example, the metrics can be rather generic: accepting a string (supported by pyReCo), a function handle that needs to follow a specific syntax, or a metric object (I think we have a metrics class in pyreco). But this is only for iterations 2+.
